src/main/kotlin/com/vanniktech/maven/publish/nexus/Nexus.kt

```diff
@@ -0,0 +1,137 @@
+import retrofit2.Retrofit
+import retrofit2.converter.moshi.MoshiConverterFactory
+import java.io.IOException
+import java.lang.IllegalArgumentException
+
+class Nexus(username: String, password: String, val groupId: String, baseUrl: String) {
+  private val service by lazy {
+    val okHttpClient = OkHttpClient.Builder()
+      .addInterceptor(NexusOkHttpInterceptor(username, password))
+      .build()
+    val retrofit = Retrofit.Builder()
+      .addConverterFactory(MoshiConverterFactory.create())
+      .client(okHttpClient)
+      .baseUrl(baseUrl)
+      .build()
+
+    retrofit.create(NexusService::class.java)
+  }
+
+  private fun getProfileRepositories(): List<Repository>? {
+    val profileRepositoriesResponse = service.getProfileRepositories().execute()
+
+    if (!profileRepositoriesResponse.isSuccessful) {
+      throw IOException("Cannot get profileRepositories: ${profileRepositoriesResponse.errorBody()?.string()}")
+    }
+
+    return profileRepositoriesResponse.body()?.data
+  }
+
+  private fun findStagingRepository(): Repository {
+    val candidateRepositories = getProfileRepositories()
+      ?.filter { it.repositoryId.startsWith(groupId.replace(".", "")) } ?: emptyList()
+
+    if (candidateRepositories.isEmpty()) {
+      throw IllegalArgumentException("No staging repository find. Did you call ./gradlew uploadArchives ?")
+    }
+
+    if (candidateRepositories.size > 1) {
+      throw IllegalArgumentException("You have ${candidateRepositories.size} staging repositories, please login on https://oss.sonatype.org and drop stale repositories. This script only works with one active staging repository at a time.")
+    }
+    return candidateRepositories[0]
+  }
+
+  fun findAndCloseStagingRepository(): String {
+    val stagingRepository = findStagingRepository()
+    val repositoryId = stagingRepository.repositoryId
+
+    if (stagingRepository.type != "open") {
+      throw IllegalArgumentException("Repository $repositoryId is of type '${stagingRepository.type}' and not 'open'")
+    }
+
+    println("Closing repository: $repositoryId")
+    val response = service.closeRepository(TransitionRepositoryInput(TransitionRepositoryInputData(listOf(repositoryId)))).execute()
+    if (!response.isSuccessful) {
+      throw IOException("Cannot close repository: ${response.errorBody()?.string()}")
+    }
+
+    waitForClose(repositoryId)
+
+    return repositoryId
+  }
+
+  private fun waitForClose(repositoryId: String) {
+
+    val startMillis = System.currentTimeMillis()
+
+    val waitingChars = listOf(
+      PROGRESS_1,
+      PROGRESS_2,
+      PROGRESS_3,
+      PROGRESS_4,
+      PROGRESS_5,
+      PROGRESS_6,
+      PROGRESS_7
+    )
+    var i = 0
+    while (true) {
+      if (System.currentTimeMillis() - startMillis > CLOSE_TIMEOUT_MILLIS) {
+        throw IOException("Timeout waiting for repository close")
+      }
+
+      print("\r${waitingChars[i++ % waitingChars.size]} waiting for close...")
```

Comment:
Does this actually work with gradle? In my experience Gradle basically hijacks stdout and ignores carriage return characters until a newline char comes though

Reply:
I think I've seen it working but it didn't look great, we should certainly remove it and replace it by something more native, if there is such a thing.

Reply:
After some hacking around, I was able to get it to work! Requires calling System.out.flush() after each time

Reply:
Please merge this stuff 🙏
I'm waiting for this pretty much to fully automate my release process :)

Reply:
I added the `System.out.flush()`